File: iceberg.py
pip install pygit2


from pyspark.sql import SparkSession
from pyspark.sql.functions import col, lit, sha2, concat_ws, coalesce
from reportlab.lib.pagesizes import letter
from reportlab.platypus import SimpleDocTemplate, Paragraph, Table, TableStyle, Spacer
from reportlab.lib.styles import getSampleStyleSheet
from reportlab.lib import colors


def main():
    # Initialize Spark session
    spark = SparkSession.builder \
        .appName("DatasetComparison") \
        .getOrCreate()

    # Input Paths
    path1 = "default.icebergtest_ac0503"  # Hive table
    path2 = "s3a://cdpmodakbucket/mt24082/dataset2.parquet/"  # Parquet path

    # Read Datasets
    df1 = spark.table(path1)
    df2 = spark.read.parquet(path2)

    # Collect schema and row counts
    schema1 = df1.schema
    schema2 = df2.schema
    count1 = df1.count()
    count2 = df2.count()

    # Normalize schema for comparison
    normalizedSchema1 = set((f.name.lower(), f.dataType) for f in schema1)
    normalizedSchema2 = set((f.name.lower(), f.dataType) for f in schema2)
    schemaChanged = normalizedSchema1 != normalizedSchema2

    # Align schemas
    allColumns = list(set(df1.columns + df2.columns))
    for colName in allColumns:
        if colName not in df1.columns:
            df1 = df1.withColumn(colName, lit(None))
        if colName not in df2.columns:
            df2 = df2.withColumn(colName, lit(None))

    df1 = df1.select(*allColumns)
    df2 = df2.select(*allColumns)

    def add_hash_column(df):
        return df.withColumn(
            "row_hash",
            sha2(concat_ws("||", *[coalesce(col(c).cast("string"), lit("NULL")) for c in allColumns]), 256)
        )

    df1Hashed = add_hash_column(df1)
    df2Hashed = add_hash_column(df2)

    # Find mismatched rows
    df1OnlyRows = df1Hashed.select("row_hash").subtract(df2Hashed.select("row_hash"))
    df2OnlyRows = df2Hashed.select("row_hash").subtract(df1Hashed.select("row_hash"))

    df1OnlyCount = df1OnlyRows.count()
    df2OnlyCount = df2OnlyRows.count()
    mismatchedCount = df1OnlyCount + df2OnlyCount

    # PDF Report Generation
    pdfPath = "/tmp/data_comparison_summary.pdf"
    doc = SimpleDocTemplate(pdfPath, pagesize=letter)
    styles = getSampleStyleSheet()
    elements = []

    elements.append(Paragraph("Data Validation Report", styles["Heading1"]))
    elements.append(Spacer(1, 12))

    # Pipeline Metadata
    elements.append(Paragraph("1. Pipeline Metadata", styles["Heading2"]))
    pipelineName = "Manual_Run"
    scheduledBy = "CLI_User"
    pipelineStartTs = "Manual_Execution"
    pipelineId = "N/A"
    elements.extend([
        Paragraph(f"Pipeline Name: {pipelineName}", styles["Normal"]),
        Paragraph(f"Scheduled By: {scheduledBy}", styles["Normal"]),
        Paragraph(f"Pipeline Start Timestamp: {pipelineStartTs}", styles["Normal"]),
        Paragraph(f"Pipeline Id: {pipelineId}", styles["Normal"]),
        Paragraph(f"Source Dataset Path: {path1}", styles["Normal"]),
        Paragraph(f"Target Dataset Path: {path2}", styles["Normal"]),
        Spacer(1, 12)
    ])

    # Data Stats
    elements.append(Paragraph("2. Data Statistics", styles["Heading2"]))
    stats_data = [
        ["No.", "Category", "Source Dataset", "Target Dataset"],
        ["1", "Row Count", str(count1), str(count2)],
        ["2", "Column Count", str(len(schema1)), str(len(schema2))],
        ["3", "Differing Rows", str(df1OnlyCount), str(df2OnlyCount)],
    ]
    stats_table = Table(stats_data)
    stats_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), colors.grey),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.whitesmoke),
        ('GRID', (0, 0), (-1, -1), 1, colors.black),
    ]))
    elements.append(stats_table)
    elements.append(Spacer(1, 12))

    # Validation Checks
    elements.append(Paragraph("3. Validation Checks", styles["Heading2"]))
    validation_data = [
        ["Check", "Status"],
        ["Schema Validation", "Failed" if schemaChanged else "Passed"],
        ["Data Validation", "Failed" if mismatchedCount > 0 else "Passed"],
        ["Row Validation", "Failed" if df1OnlyCount != df2OnlyCount else "Passed"],
        ["Column Validation", "Failed" if len(schema1) != len(schema2) else "Passed"],
    ]
    validation_table = Table(validation_data)
    validation_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), colors.grey),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.whitesmoke),
        ('GRID', (0, 0), (-1, -1), 1, colors.black),
    ]))
    elements.append(validation_table)
    elements.append(Spacer(1, 12))

    # Dataset Fields
    elements.append(Paragraph("4. Dataset Fields", styles["Heading2"]))
    fields_data = [
        ["Dataset", "Fields"],
        ["Source Dataset", ", ".join(f.name for f in schema1)],
        ["Target Dataset", ", ".join(f.name for f in schema2)],
    ]
    fields_table = Table(fields_data)
    fields_table.setStyle(TableStyle([
        ('BACKGROUND', (0, 0), (-1, 0), colors.grey),
        ('TEXTCOLOR', (0, 0), (-1, 0), colors.whitesmoke),
        ('GRID', (0, 0), (-1, -1), 1, colors.black),
    ]))
    elements.append(fields_table)

    doc.build(elements)
    print(f"\n✅ PDF Report generated at: {pdfPath}\n")

    spark.stop()

if __name__ == "__main__":
    main()

File: iceberg.scala
  import org.apache.spark.sql.functions._
  import org.apache.spark.sql.SparkSession
  import java.io.FileOutputStream
  import com.itextpdf.kernel.pdf._
  import com.itextpdf.layout._
  import com.itextpdf.layout.element._
  import spark.implicits._
  
  val path1 = "default.icebergtest_ac0503"
  val path2 ="s3a://cdpmodakbucket/mt24082/dataset2.parquet/"

  val df1 = spark.table(path1)
  val df2 = spark.read.parquet(path2)

  val schema1 = df1.schema
  val schema2 = df2.schema

  val count1 = df1.count()
  val count2 = df2.count()

val normalizedSchema1 = schema1.map(f => (f.name.toLowerCase, f.dataType)).toSet
val normalizedSchema2 = schema2.map(f => (f.name.toLowerCase, f.dataType)).toSet

val schemaChanged = normalizedSchema1 != normalizedSchema2

  val allColumns = (df1.columns ++ df2.columns).distinct

  val df1Aligned = allColumns.foldLeft(df1)((df, colName) =>
    if (df.columns.contains(colName)) df else df.withColumn(colName, lit(null))
  ).select(allColumns.map(col): _*)

  val df2Aligned = allColumns.foldLeft(df2)((df, colName) =>
    if (df.columns.contains(colName)) df else df.withColumn(colName, lit(null))
  ).select(allColumns.map(col): _*)

  val df1Compare = df1Aligned.withColumn("row_hash", sha2(concat_ws("||", allColumns.map(c => coalesce(col(c).cast("string"), lit("NULL"))): _*), 256))
  val df2Compare = df2Aligned.withColumn("row_hash", sha2(concat_ws("||", allColumns.map(c => coalesce(col(c).cast("string"), lit("NULL"))): _*), 256))

  val df1OnlyRows = df1Compare.select("row_hash").except(df2Compare.select("row_hash"))
  val df2OnlyRows = df2Compare.select("row_hash").except(df1Compare.select("row_hash"))

  val df1OnlyCount = df1OnlyRows.count()
  val df2OnlyCount = df2OnlyRows.count()
  val mismatchedCount = df1OnlyCount + df2OnlyCount

  val pdfPath = "/home/spark-srv-account/data_comparison_summary.pdf_82_iceberg"
  val pdfWriter = new PdfWriter(new FileOutputStream(pdfPath))
  val pdfDoc = new PdfDocument(pdfWriter)
  val document = new Document(pdfDoc)

  document.add(new Paragraph("Data Validation Report").setBold().setFontSize(18))
  document.add(new Paragraph("\n"))

  document.add(new Paragraph("1. Pipeline Metadata").setBold().setFontSize(14))

  val pipelineName = Option(variablesMap.get("pipeline_name")).map(_.toString).getOrElse("unknown_pipeline")
  val scheduledBy = Option(variablesMap.get("job_schedule_user_id")).map(_.toString).getOrElse("unknown_user")
  val pipelineStartTs = Option(variablesMap.get("pipeline_start_ts")).map(_.toString).getOrElse("N/A")
  val pipelineId = Option(variablesMap.get("pipeline_id")).map(_.toString).getOrElse("N/A")
  val sourceDatasetPath = path1
  val targetDatasetPath = path2

  document.add(new Paragraph(s"Pipeline Name: $pipelineName"))
  document.add(new Paragraph(s"Scheduled By: $scheduledBy"))
  document.add(new Paragraph(s"Pipeline Start Timestamp: $pipelineStartTs"))
  document.add(new Paragraph(s"Pipeline Id: $pipelineId"))
  document.add(new Paragraph(s"Source Dataset Path: $sourceDatasetPath"))
  document.add(new Paragraph(s"Target Dataset Path: $targetDatasetPath"))
  document.add(new Paragraph("\n"))

  document.add(new Paragraph("2. Data Statistics").setBold().setFontSize(14))

  val statsTable = new Table(4)
  statsTable.setAutoLayout() // iText 8 style for responsive width

  statsTable.addHeaderCell(new Cell().add(new Paragraph("No.")))
  statsTable.addHeaderCell(new Cell().add(new Paragraph("Category")))
  statsTable.addHeaderCell(new Cell().add(new Paragraph("Source Dataset")))
  statsTable.addHeaderCell(new Cell().add(new Paragraph("Target Dataset")))

  statsTable.addCell(new Cell().add(new Paragraph("1")))
  statsTable.addCell(new Cell().add(new Paragraph("Row Count")))
  statsTable.addCell(new Cell().add(new Paragraph(count1.toString)))
  statsTable.addCell(new Cell().add(new Paragraph(count2.toString)))

  statsTable.addCell(new Cell().add(new Paragraph("2")))
  statsTable.addCell(new Cell().add(new Paragraph("Column Count")))
  statsTable.addCell(new Cell().add(new Paragraph(schema1.fields.length.toString)))
  statsTable.addCell(new Cell().add(new Paragraph(schema2.fields.length.toString)))

  statsTable.addCell(new Cell().add(new Paragraph("3")))
  statsTable.addCell(new Cell().add(new Paragraph("Differing Rows")))
  statsTable.addCell(new Cell().add(new Paragraph(df1OnlyCount.toString)))
  statsTable.addCell(new Cell().add(new Paragraph(df2OnlyCount.toString)))

  document.add(statsTable)
  document.add(new Paragraph("\n"))

  document.add(new Paragraph("3. Validation Checks").setBold().setFontSize(14))

  val validationTable = new Table(2)
  validationTable.setAutoLayout()

  validationTable.addHeaderCell(new Cell().add(new Paragraph("Check")))
  validationTable.addHeaderCell(new Cell().add(new Paragraph("Status")))

  validationTable.addCell(new Cell().add(new Paragraph("Schema Validation")))
  validationTable.addCell(new Cell().add(new Paragraph(if (schemaChanged) "Failed" else "Passed")))

  validationTable.addCell(new Cell().add(new Paragraph("Data Validation")))
  validationTable.addCell(new Cell().add(new Paragraph(if (mismatchedCount > 0) "Failed" else "Passed")))

  validationTable.addCell(new Cell().add(new Paragraph("Row Validation")))
  validationTable.addCell(new Cell().add(new Paragraph(if (df1OnlyCount != df2OnlyCount) "Failed" else "Passed")))

  validationTable.addCell(new Cell().add(new Paragraph("Column Validation")))
  validationTable.addCell(new Cell().add(new Paragraph(if (schema1.length != schema2.length) "Failed" else "Passed")))

  document.add(validationTable)
  document.add(new Paragraph("\n"))

  document.add(new Paragraph("4. Dataset Fields").setBold().setFontSize(14))

  val fieldsTable = new Table(2)
  fieldsTable.setAutoLayout()

  fieldsTable.addHeaderCell(new Cell().add(new Paragraph("Dataset")))
  fieldsTable.addHeaderCell(new Cell().add(new Paragraph("Fields")))

  fieldsTable.addCell(new Cell().add(new Paragraph("Source Dataset")))
  fieldsTable.addCell(new Cell().add(new Paragraph(schema1.map(_.name).mkString(", "))))

  fieldsTable.addCell(new Cell().add(new Paragraph("Target Dataset")))
  fieldsTable.addCell(new Cell().add(new Paragraph(schema2.map(_.name).mkString(", "))))

  document.add(fieldsTable)
  document.close()

  println(s"PDF Report generated: $pdfPath")

